ios/browser/ui/webui/public/brave_web_ui_ios_data_source.mm

```diff
@@ -0,0 +1,309 @@
+                                        const std::string& value) {
+  localized_strings_.Set(name, value);
+  replacements_[name] = value;
+}
+
+void BraveWebUIIOSDataSource::AddLocalizedString(const std::string& name,
+                                                 int ids) {
+  localized_strings_.Set(name, web::GetWebClient()->GetLocalizedString(ids));
+  replacements_[name] =
+      base::UTF16ToUTF8(web::GetWebClient()->GetLocalizedString(ids));
+}
+
+void BraveWebUIIOSDataSource::AddLocalizedStrings(
+    const base::Value::Dict& localized_strings) {
+  localized_strings_.Merge(localized_strings.Clone());
+  ui::TemplateReplacementsFromDictionaryValue(localized_strings,
+                                              &replacements_);
+}
+
+void BraveWebUIIOSDataSource::AddLocalizedStrings(
+    base::span<const webui::LocalizedString> strings) {
+  for (const auto& str : strings) {
+    AddLocalizedString(str.name, str.id);
+  }
+}
+
+void BraveWebUIIOSDataSource::AddBoolean(const std::string& name, bool value) {
+  localized_strings_.Set(name, value);
+}
+
+void BraveWebUIIOSDataSource::UseStringsJs() {
+  use_strings_js_ = true;
+}
+
+void BraveWebUIIOSDataSource::EnableReplaceI18nInJS() {
+  should_replace_i18n_in_js_ = true;
+}
+
+bool BraveWebUIIOSDataSource::ShouldReplaceI18nInJS() const {
+  return should_replace_i18n_in_js_;
+}
+
+void BraveWebUIIOSDataSource::AddResourcePath(const std::string& path,
+                                              int resource_id) {
+  path_to_idr_map_[path] = resource_id;
+}
+
+void BraveWebUIIOSDataSource::AddResourcePaths(
+    base::span<const webui::ResourcePath> paths) {
+  for (const auto& path : paths) {
+    AddResourcePath(path.path, path.id);
+  }
+}
+
+void BraveWebUIIOSDataSource::SetDefaultResource(int resource_id) {
+  default_resource_ = resource_id;
+}
+
+void BraveWebUIIOSDataSource::DisableDenyXFrameOptions() {
+  deny_xframe_options_ = false;
+}
+
+const ui::TemplateReplacements* BraveWebUIIOSDataSource::GetReplacements()
+    const {
+  return &replacements_;
+}
+
+// URLDataSourceIOS implementation:
+
+std::string BraveWebUIIOSDataSource::GetSource() const {
+  return base::StrCat({kChromeUIScheme, url::kStandardSchemeSeparator});
+}
+
+void BraveWebUIIOSDataSource::StartDataRequest(
+    const std::string& path,
+    web::URLDataSourceIOS::GotDataCallback callback) {
+  EnsureLoadTimeDataDefaultsAdded();
+
+  if (use_strings_js_) {
+    bool from_js_module = path == "strings.m.js";
+    if (from_js_module || path == "strings.js") {
+      SendLocalizedStringsAsJSON(std::move(callback), from_js_module);
+      return;
+    }
+  }
+
+  int resource_id = PathToIdrOrDefault(path);
+  DCHECK_NE(resource_id, -1);
+  scoped_refptr<base::RefCountedMemory> response(
+      web::GetWebClient()->GetDataResourceBytes(resource_id));
+  std::move(callback).Run(response);
+}
+
+std::string BraveWebUIIOSDataSource::GetMimeType(
```

Comment:
why aren't we subclassing `WebUIDataSourceImpl` to get this stuff? We don't want to duplicate it

Reply:
per dm I think we can have `BraveWebUIIOSDataSource` own an instance of `WebUIIOSDataSourceImpl` that is created through `WebUIIOSDataSource::Create` and then proxy these methods to it